auth/authenticator.go

```diff
@@ -0,0 +1,225 @@
+// © Copyright IBM Corporation 2020. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/IBM/go-sdk-core/v4/core"
+)
+
+const (
+	AUTHTYPE_COUCHDB_SESSION = "couchdbSession"
+)
+
+var requestSessionMutex sync.Mutex
+
+// CouchDbSessionAuthenticator uses username and password to obtain
+// CouchDB authentication cookie, and adds the cookie to requests.
+type CouchDbSessionAuthenticator struct {
+	URL, Username, Password string
+	Client                  *http.Client
```

Comment:
We had this conversation on Slack about this and agreed between @emlaver and @ricellis that having public `Client` var to allow setting custom `http.Client` is sufficient.

Here is my write-up for possible approaches and cons and pros for this, it includes view on `SetHTTPClient` https://gist.github.com/eiri/f9e6a2a7ed7e0cc6bdf7a8fec76c387b

Reply:
In case of IAM you need to ask token from a Token Server `https://iam.cloud.ibm.com/identity/token`, so you need to open a new channel in the network towards this server.

But with CouchdbSessionAuthenticator you connect to Cloudant Server, towards an already opened direction in the network. For this server a Client is already defined, and maybe some connections already built up and those could be used for the authentication.
Also different transport settings could be used in the client I found here a good list about the settings: https://golang.org/pkg/net/http/#RoundTripper

And which option will be implemented from the listed 3 variations?

Reply:
> But with CouchdbSessionAuthenticator you connect to Cloudant Server, towards an already opened direction in the network. 

Not really, this is authenticator, so it called first, before any connection opened and then it holds its own client with transport, that does all the things you mentioned. If one needs custom transport it can be set on authenticator's client directly, same way as for IAM, that's why `Client` var is public.

Reply:
What I also found, that the SSL verification settings is stored in the Client: https://github.com/IBM/go-sdk-core/blob/775201f679d69afb97f3259ababb24d482089410/v4/core/base_service.go#L184
What is the alternate solution to transfer this setting to the authenticator?

Reply:
According the current implementation there is two client used, one for the server and an other one to the same server. There will be 2 network tunnels opened to the same address.

Reply:
Yes, that's right, I thought we already established that. What about it?

Reply:
I cannot see where these transport options are transferred. What code should be created by the user?

Reply:
@vmatyusGitHub I was busy with other stuff, give me some time to finish with url changes here that you requested and I'll show you the example, ok?

Reply:
Yeah, just take your time.

Reply:
I found an example in t your gist: https://gist.github.com/eiri/f9e6a2a7ed7e0cc6bdf7a8fec76c387b#just-use-public-field-client
Did you noticed there is a CookieJar in the Client: https://golang.org/pkg/net/http/#Client ?

Reply:
Yes, I'm familiar with standard golang's packages